### PR TITLE
Feature/player registration for ball scoring

### LINKED ISSUE #76: [Hoops and pegs] Player Registration for Ball Passing Through Correct Hoop
Task for #42 

# Implement Player Registration for Ball Passing Through Correct Hoop

## Description:

In our game, a player scores a point when their ball passes through the correct hoop. We need to implement a feature that registers this event and updates the player's score accordingly.

## Acceptance Criteria:

Update the `Ball`  or `player` class to include a method that detects when the ball has passed through a hoop.
Implement a mechanism in the Player class to register when their ball has passed through the correct hoop.
Update the player's score when their ball passes through the correct hoop.
Write unit tests to verify that the score updates correctly when the ball passes through the correct hoop.

### PR DESCRIPTION
**Pull request by Oskmey**
# Feature/player registration for ball scoring
## Description

Scoring based on direction: The ability to score points on hoops has been improved. This is achieved with the help of the northSouthCollisionDirection and eastWestCollisionDirection methods.

Addition of pegs: Pegs have been added to the game. These new elements have their own collision and textures.

Increased hoop size: The size of the hoops has been increased to make scoring easier. The scoring width is now 32, up from 22, which should make it easier for players to score points.

Hoop hitboxes: New hitboxes have been added to the hoops. These include left and right post hitboxes. New methods have been implemented to create these hitboxes based on the direction of the hoop.

Interface comments: Added some java interface comments 

### Method: `northSouthCollisionDirection`
This method determines the direction of a collision between the ball and a hoop along the East-West axis. It leverages the fact that the Entity class now stores its ***previous position***. Even though the Entity class does not have a velocity attribute, we can infer if an entity has changed its position by comparing its current and previous positions.

### Method: `eastWestCollisionDirection`

 This method determines the direction of a collision between the ball and a hoop along the East-West axis. It leverages the fact that the Entity class now stores its ***previous position***. Even though the Entity class does not have a velocity attribute, we can infer if an entity has changed its position by comparing its current and previous positions.

Closes #76 (I hoope lol)
Closes #78 

> [!IMPORTANT]
> Model can create rotated drawable entities ( `Entity` does not have rotation). 
> Player cannot win.
> Player subscribes to `EventPublisher` in the `Model` class this might be weird

### CODE REVIEW

File: crocket/src/main/java/com/crocket/controller/CroquetController.java [hunk 1]
```diff
@@ -74,6 +74,7 @@ public void keyTyped(KeyEvent ke) {
 
     }
 
+
```
Comment (filipkauffeldt) [inline crocket/src/main/java/com/crocket/controller/CroquetController.java:77]: Remove new line

File: crocket/src/main/java/com/crocket/model/Player.java [hunk 2]
```diff
@@ -48,6 +49,8 @@ public void handleEvent(PassTargetEvent event) {
         Entity target = event.getTarget();
         if (this.remainingTargets.peek() == target) {
             this.remainingTargets.poll();
+            this.incrementScore();
```
Comment (filipkauffeldt) [inline crocket/src/main/java/com/crocket/model/Player.java:52]: What is score used for? Are we not just supposed to pass under them?

File: crocket/src/main/java/com/crocket/model/Player.java [hunk 3]
```diff
@@ -64,6 +67,13 @@ public void applyPowerUp(Integer index) {
         this.powerUps.remove(powerUp);
     } 
 
+
```
Comment (filipkauffeldt) [inline crocket/src/main/java/com/crocket/model/Player.java:70]: Remove new lines

File: crocket/src/main/java/com/crocket/model/entity/CollisionHandler.java [hunk 4]
```diff
@@ -30,50 +31,60 @@ static boolean intersect(Hitbox hitbox1, Hitbox hitbox2) {
         int h2Width = hitbox2.getWidth();
         int h2Height = hitbox2.getHeight();
 
-        boolean xMax1IsGreaterThanxPos2 = h1X + h1Width > h2X;
+        return (h1X + h1Width > h2X &&
```
Comment (filipkauffeldt) [inline crocket/src/main/java/com/crocket/model/entity/CollisionHandler.java:34]: Does this work better than the previous solution?

File: crocket/src/main/java/com/crocket/model/interfaces/IEventListener.java [hunk 5]
```diff
@@ -3,8 +3,19 @@
...
+ * 
+ * This interface is expected to be implemented by classes that need to respond to target events.
+ */
 public interface IEventListener {
     public void handleEvent(PassTargetEvent event);
     
     public void handleEvent(HitPowerUpEvent event);
-}
+
+
```
Comment (filipkauffeldt) [inline crocket/src/main/java/com/crocket/model/interfaces/IEventListener.java:20]: Remove new lines

File: crocket/src/main/java/com/crocket/model/interfaces/ITargetEvent.java [hunk 6]
```diff
@@ -0,0 +1,10 @@
+package com.crocket.model.interfaces;
+
+import com.crocket.model.entity.Ball;
+import com.crocket.model.entity.Entity;
+
+public interface ITargetEvent {
+    public Entity getTarget();
```
Comment (filipkauffeldt) [inline crocket/src/main/java/com/crocket/model/interfaces/ITargetEvent.java:7]: Should this maybe be a separate type `Target`?

**Review by filipkauffeldt — COMMENTED** [review]
Looks good for the most part

File: crocket/src/main/java/com/crocket/model/interfaces/ITargetEvent.java (hunk 6 above)
Comment (Oskmey) [reply crocket/src/main/java/com/crocket/model/interfaces/ITargetEvent.java]: Removed the file

File: crocket/src/main/java/com/crocket/model/interfaces/IEventListener.java (hunk 5 above)
Comment (Oskmey) [reply crocket/src/main/java/com/crocket/model/interfaces/IEventListener.java]: removed lines

File: crocket/src/main/java/com/crocket/model/entity/CollisionHandler.java (hunk 4 above)
Comment (Oskmey) [reply crocket/src/main/java/com/crocket/model/entity/CollisionHandler.java:34]: Pontus mentioned that the code behaves oddly with small hitboxes, such as those found in post hitboxes within the hoop.

File: crocket/src/main/java/com/crocket/model/Player.java (hunk 2 above)
Comment (Oskmey) [reply crocket/src/main/java/com/crocket/model/Player.java:52]: Will fix

File: crocket/src/main/java/com/crocket/model/Player.java (hunk 3 above)
Comment (Oskmey) [reply crocket/src/main/java/com/crocket/model/Player.java]: Fixed

File: crocket/src/main/java/com/crocket/controller/CroquetController.java (hunk 1 above)
Comment (Oskmey) [reply crocket/src/main/java/com/crocket/controller/CroquetController.java]: Fixed

File: crocket/src/main/java/com/crocket/model/Level1.java [hunk 7]
```diff
@@ -20,6 +21,9 @@ public class Level1 implements ILevel {
     private Set<Entity> entities = new HashSet<Entity>();
     private Set<IMovable> movables = new HashSet<IMovable>();
     private Set<ICollidable> collidables = new HashSet<ICollidable>();
+    private Set<Entity> hoops = new HashSet<Entity>();
+    private int startXBallPos; 
+    private int startYBallPos;
```
Comment (Pontus-Wikstrom) [inline crocket/src/main/java/com/crocket/model/Level1.java:26]: Are these ever used?

**Review by Pontus-Wikstrom — APPROVED** [review]
LGTM, check comments, otherwise try to have smaller PR's in the future

**Review by filipkauffeldt — APPROVED** [review]

**Pull request merged**